Add is_tensorlist_like_type helper

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #123454
* #123453
* #123437
* #123129
* #123110
* __->__ #123109
* #123108

Checks if the type of an argument in a schema is some form of
TensorList.

Test Plan:
- new test

zou3519 committed:
Add is_tensorlist_like_type helper

Checks if the type of an argument in a schema is some form of
TensorList.

Test Plan:
- new test

[ghstack-poisoned]

zou3519 committed:
Update on "Add is_tensorlist_like_type helper"


Checks if the type of an argument in a schema is some form of
TensorList.

Test Plan:
- new test

[ghstack-poisoned]

albanD commented on test/test_custom_ops.py:
```diff
@@ -1730,6 +1730,28 @@ def test_is_functional_schema(self):
+            torch.ops.aten.index.Tensor._schema.arguments[1].type,
+            # Tensor[]?
+            torch._C.parse_schema("foo(Tensor[]? x) -> ()").arguments[0].type,
+            # Tensor?[]?
```
Are these actually valid schemas??

zou3519 replied:
It's unclear to me?

albanD replied:
I'm surprised that parse_schema doesn't raise....

zou3519 replied:
Yes, these are all valid (and used)

albanD commented on torch/_library/utils.py:
```diff
@@ -103,6 +104,15 @@ def is_functional(schema):
     return is_functional(schema)
 
 
+def is_tensorlist_like_type(typ: torch.Type):
```
I guess we have this in torchgen already? Do we expect deduplication to be able to happen between the two or they can't depend on each other?

albanD replied:
Ho? Which ones don't?

zou3519 replied:
resolved offline: long-term state is that we want deduplication but some internal builds make it difficult to deduplicate.

zou3519 committed:
Update on "Add is_tensorlist_like_type helper"


Checks if the type of an argument in a schema is some form of
TensorList.

Test Plan:
- new test

[ghstack-poisoned]

zou3519 committed:
Update on "Add is_tensorlist_like_type helper"


Checks if the type of an argument in a schema is some form of
TensorList.

Test Plan:
- new test

[ghstack-poisoned]

zou3519 committed:
Update on "Add is_tensorlist_like_type helper"


Checks if the type of an argument in a schema is some form of
TensorList.

Test Plan:
- new test

[ghstack-poisoned]

zou3519 committed:
Update on "Add is_tensorlist_like_type helper"


Checks if the type of an argument in a schema is some form of
TensorList.

Test Plan:
- new test

[ghstack-poisoned]

albanD reviewed (approved):
Ok!

zou3519 committed:
Update on "Add is_tensorlist_like_type helper"


Checks if the type of an argument in a schema is some form of
TensorList.

Test Plan:
- new test

[ghstack-poisoned]

zou3519 committed:
Update on "Add is_tensorlist_like_type helper"


Checks if the type of an argument in a schema is some form of
TensorList.

Test Plan:
- new test

[ghstack-poisoned]
